File: PaperPiano/js/hand-tracker.js
/* =========================================================
   HandTracker — MediaPipe Hands wrapper
   =========================================================
   Tracks up to 2 hands using the MediaPipe Hands
   *Solutions* API loaded from CDN.  Exposes only EXTENDED
   fingertip positions (curled fingers are ignored).

   Includes:
     • Finger-curl detection (only report extended fingers)
     • Exponential-moving-average smoothing on landmarks
     • Grace period — keeps last-known hand for a few frames
       when MediaPipe briefly drops detection
   ========================================================= */
'use strict';

class HandTracker {
    constructor () {
        /** @type {Hands|null} */
        this.hands      = null;
        this.results    = null;
        this.ready      = false;
        this.processing = false;
        /** External callback — set by main.js */
        this.onResults  = null;

        /* ---- smoothing state ---- */
        /** Smoothed landmarks per hand index: Map<handIndex, landmark array> */
        this._smoothed  = new Map();
        /** EMA factor: 0 = ignore new data, 1 = no smoothing */
        this._alpha     = 0.45;

        /* ---- grace period ---- */
        /** How many consecutive empty frames before we drop a hand */
        this._graceFrames = 5;
        /** Counter of frames since we last saw hands */
        this._missCount   = 0;
        /** Cached last valid results (for grace period) */
        this._lastGoodResults = null;
    }

    /* ---------- lifecycle ---------- */

    async init () {
        // eslint-disable-next-line no-undef
        this.hands = new Hands({
            locateFile: file =>
                `https://cdn.jsdelivr.net/npm/@mediapipe/hands/${file}`
        });

        this.hands.setOptions({
            maxNumHands:            2,
            modelComplexity:        1,      // 1 = Full (more accurate, less jitter)
            minDetectionConfidence: 0.5,
            minTrackingConfidence:  0.45,
        });

        this.hands.onResults(r => {
            this._processResults(r);
            this.processing = false;
            if (this.onResults) this.onResults(this.results);
        });

        await this.hands.initialize();
        this.ready = true;
    }

    /** Send a video frame for processing. */
    async send (videoEl) {
        if (!this.ready || this.processing) return;
        this.processing = true;
        try {
            await this.hands.send({ image: videoEl });
        } catch (e) {
            this.processing = false;
        }
    }

    /* ---------- internal: smoothing + grace ---------- */

    _processResults (raw) {
        const hasHands = raw.multiHandLandmarks && raw.multiHandLandmarks.length > 0;

        if (hasHands) {
            this._missCount = 0;
            this._lastGoodResults = raw;

            // Apply EMA smoothing to each hand's landmarks
            const smoothedLandmarks = raw.multiHandLandmarks.map((lm, hi) => {
                return this._smoothLandmarks(hi, lm);
            });

            this.results = {
                multiHandLandmarks:  smoothedLandmarks,
                multiHandedness:     raw.multiHandedness,
                image:               raw.image,
            };
        } else {
            this._missCount++;

            if (this._missCount <= this._graceFrames && this._lastGoodResults) {
                // Keep showing last known hand positions during brief drops
                this.results = this._lastGoodResults;
            } else {
                // Truly lost — clear everything
                this.results = raw;
                this._smoothed.clear();
                this._lastGoodResults = null;
            }
        }
    }

    /**
     * Exponential moving average on landmark positions.
     * Reduces jitter while keeping responsiveness.
     */
    _smoothLandmarks (handIndex, rawLandmarks) {
        const prev = this._smoothed.get(handIndex);

        if (!prev || prev.length !== rawLandmarks.length) {
            // First frame for this hand — just store raw
            const copy = rawLandmarks.map(p => ({ x: p.x, y: p.y, z: p.z }));
            this._smoothed.set(handIndex, copy);
            return copy;
        }

        const smoothed = rawLandmarks.map((p, i) => ({
            x: prev[i].x + this._alpha * (p.x - prev[i].x),
            y: prev[i].y + this._alpha * (p.y - prev[i].y),
            z: prev[i].z + this._alpha * (p.z - prev[i].z),
        }));

        this._smoothed.set(handIndex, smoothed);
        return smoothed;
    }

    /* ---------- queries ---------- */

    /**
     * Returns only EXTENDED fingertips currently visible.
     * Curled/bent fingers are excluded to prevent accidental triggers.
     *
     * A finger is considered extended if its tip is farther from the
     * wrist than its MCP (knuckle) joint — a simple but effective test.
     *
     * Returns: { x, y, z, finger, hand }
     */
    getFingerTips () {
        if (!this.results || !this.results.multiHandLandmarks) return [];

        /*
         * Landmark indices for each finger:
         *   tip, dip, pip, mcp
         *   Thumb:  4, 3, 2, 1   (special: compare tip.x vs mcp.x for abduction)
         *   Index:  8, 7, 6, 5
         *   Middle: 12,11,10, 9
         *   Ring:   16,15,14,13
         *   Pinky:  20,19,18,17
         */
        const FINGERS = [
            { name: 'index',  tip: 8,  dip: 7,  pip: 6,  mcp: 5  },
            { name: 'middle', tip: 12, dip: 11, pip: 10, mcp: 9  },
            { name: 'ring',   tip: 16, dip: 15, pip: 14, mcp: 13 },
            { name: 'pinky',  tip: 20, dip: 19, pip: 18, mcp: 17 },
        ];

        const tips = [];
        this.results.multiHandLandmarks.forEach((lm, hi) => {
            const wrist = lm[0];

            for (const f of FINGERS) {
                // Extended test: tip must be farther from wrist than pip joint
                // (works regardless of hand orientation)
                const tipDist = Math.hypot(lm[f.tip].x - wrist.x,
                                           lm[f.tip].y - wrist.y);
                const pipDist = Math.hypot(lm[f.pip].x - wrist.x,
                                           lm[f.pip].y - wrist.y);
                if (tipDist <= pipDist) continue; // finger is curled

                tips.push({
                    x:      lm[f.tip].x,
                    y:      lm[f.tip].y,
                    z:      lm[f.tip].z,
                    finger: f.name,
                    hand:   hi,
                });
            }
        });
        return tips;
    }

    /**
     * Returns the full set of 21 landmarks per hand (for debug drawing).
     */
    getLandmarks () {
        if (!this.results) return [];
        return this.results.multiHandLandmarks || [];
    }
}

window.HandTracker = HandTracker;

File: PaperPiano/hand_tracker.py
"""
HandTracker — Real-time hand landmark detection using MediaPipe Tasks Vision.
Tracks fingertip positions and detects tap gestures for top-down camera setups.

Tap detection works by tracking each fingertip's z-coordinate (depth) over time.
A tap is registered when a finger moves downward (z increases) past a threshold
and the fingertip is over a target area. Uses a per-finger state machine:
  UP → PRESSING (z moving down) → DOWN (tap triggered) → UP (z moving back up)
"""

import os
import time
import urllib.request

import mediapipe as mp
from mediapipe.tasks.python import BaseOptions
from mediapipe.tasks.python.vision import (
    HandLandmarker,
    HandLandmarkerOptions,
    RunningMode,
)


# Fingertip landmark indices: thumb=4, index=8, middle=12, ring=16, pinky=20
TIP_IDS = [4, 8, 12, 16, 20]

# Hand skeleton connections for drawing
HAND_CONNECTIONS = [
    (0, 1), (1, 2), (2, 3), (3, 4),           # thumb
    (0, 5), (5, 6), (6, 7), (7, 8),           # index
    (5, 9), (9, 10), (10, 11), (11, 12),      # middle
    (9, 13), (13, 14), (14, 15), (15, 16),    # ring
    (13, 17), (17, 18), (18, 19), (19, 20),   # pinky
    (0, 17),                                    # palm base
]

MODEL_URL = (
    'https://storage.googleapis.com/mediapipe-models/'
    'hand_landmarker/hand_landmarker/float16/1/hand_landmarker.task'
)
MODEL_PATH = os.path.join(os.path.dirname(__file__), 'hand_landmarker.task')


class HandTracker:
    def __init__(self):
        self.landmarker: HandLandmarker | None = None
        self.ready = False
        self._last_result = None
        self._timestamp_ms = 0

        # Tap detection state
        self._finger_z: dict[tuple, tuple[float, float]] = {}  # (hand, tip) → (z, time)
        self._tap_cooldown: dict[tuple, float] = {}  # (hand, tip) → last_tap_time
        self._last_cleanup = 0.0

    def init(self):
        """Initialize MediaPipe HandLandmarker (Tasks API)."""
        try:
            # Download model if not cached locally
            if not os.path.exists(MODEL_PATH):
                print('Downloading hand-tracking model…')
                urllib.request.urlretrieve(MODEL_URL, MODEL_PATH)
                print('Model downloaded ✓')

            options = HandLandmarkerOptions(
                base_options=BaseOptions(model_asset_path=MODEL_PATH),
                running_mode=RunningMode.VIDEO,
                num_hands=2,
                min_hand_detection_confidence=0.5,
                min_hand_presence_confidence=0.5,
                min_tracking_confidence=0.5,
            )
            self.landmarker = HandLandmarker.create_from_options(options)
            self.ready = True
            print('Hand tracking ready ✓')
        except Exception as e:
            print(f'HandTracker init failed: {e}')
            self.ready = False

    def detect(self, frame_rgb):
        """
        Run detection on an RGB frame (numpy array).
        Returns a HandLandmarkerResult or None.
        """
        if not self.ready or not self.landmarker:
            return None
        try:
            # MediaPipe Tasks requires monotonically increasing timestamps
            self._timestamp_ms += 33  # ~30 fps
            mp_image = mp.Image(image_format=mp.ImageFormat.SRGB, data=frame_rgb)
            result = self.landmarker.detect_for_video(mp_image, self._timestamp_ms)
            return result
        except Exception:
            return None

    def get_finger_tips(self, results, frame_width: int, frame_height: int) -> list[dict]:
        """
        Extract all fingertip positions from hand results.
        Returns a list of hand dicts, each with 'tips' and 'landmarks'.
        """
        if not results or not results.hand_landmarks:
            return []

        hands = []
        for hand_lms in results.hand_landmarks:
            tips = []
            for tid in TIP_IDS:
                lm = hand_lms[tid]
                tips.append({
                    'x': lm.x * frame_width,
                    'y': lm.y * frame_height,
                    'z': lm.z,
                })

            landmarks = []
            for lm in hand_lms:
                landmarks.append({
                    'x': lm.x * frame_width,
                    'y': lm.y * frame_height,
                    'z': lm.z,
                })

            hands.append({
                'x': hand_lms[8].x * frame_width,   # index finger primary
                'y': hand_lms[8].y * frame_height,
                'tips': tips,
                'landmarks': landmarks,
            })

        return hands

    # ── Tap detection ───────────────────────────────────────────────────────
    # For top-down camera: detect finger taps by checking if a finger is
    # extended and pressing downward (tip below its PIP joint = curling down).
    # We use a per-finger state machine with cooldown.

    def get_tapping_tips(self, results, frame_width: int,
                         frame_height: int) -> list[dict]:
        """
        Return only the fingertips that are currently in a 'tap' state.
        A tap = the finger is extended and pressing down (tip is far from
        the wrist relative to MCP, AND the finger is reasonably straight).

        For a top-down camera, we detect taps by checking:
          1. Finger is extended (tip further from wrist than PIP in image-y)
          2. Fingertip moved downward recently (z-delta tracking)
          3. Cooldown prevents re-triggering

        Returns list of dicts with 'x', 'y', 'finger_id', 'hand_idx'.
        """
        if not results or not results.hand_landmarks:
            return []

        now = time.time()
        taps = []

        # PIP landmark indices for each finger (used to check extension)
        # thumb=4→IP=3, index=8→PIP=6, middle=12→PIP=10, ring=16→PIP=14, pinky=20→PIP=18
        PIP_IDS = [3, 6, 10, 14, 18]
        MCP_IDS = [2, 5, 9, 13, 17]

        for hand_idx, hand_lms in enumerate(results.hand_landmarks):
            for i, (tip_id, pip_id, mcp_id) in enumerate(
                zip(TIP_IDS, PIP_IDS, MCP_IDS)
            ):
                tip = hand_lms[tip_id]
                pip = hand_lms[pip_id]
                mcp = hand_lms[mcp_id]

                tip_x = tip.x * frame_width
                tip_y = tip.y * frame_height

                # Check if finger is extended:
                # For index/middle/ring/pinky: tip.y > pip.y means pointing
                # down in image space (camera facing down on table).
                # For thumb: use x-distance instead.
                if i == 0:  # thumb
                    extended = abs(tip.x - mcp.x) > abs(pip.x - mcp.x)
                else:
                    # Finger is extended if tip is further from MCP than PIP
                    # in the image y-axis (downward = toward table)
                    tip_dist = ((tip.x - mcp.x)**2 + (tip.y - mcp.y)**2) ** 0.5
                    pip_dist = ((pip.x - mcp.x)**2 + (pip.y - mcp.y)**2) ** 0.5
                    extended = tip_dist > pip_dist * 0.85

                if not extended:
                    continue

                # Z-based tap: track z over time per finger
                fkey = (hand_idx, tip_id)
                prev = self._finger_z.get(fkey)
                cur_z = tip.z

                if prev is not None:
                    prev_z, prev_time = prev
                    z_delta = cur_z - prev_z  # positive = moving toward table

                    # Tap trigger: z moved down enough AND cooldown passed
                    cooldown_key = fkey
                    last_tap = self._tap_cooldown.get(cooldown_key, 0)

                    if z_delta > 0.008 and (now - last_tap) > 0.25:
                        taps.append({
                            'x': tip_x,
                            'y': tip_y,
                            'finger_id': tip_id,
                            'hand_idx': hand_idx,
                        })
                        self._tap_cooldown[cooldown_key] = now

                self._finger_z[fkey] = (cur_z, now)

        # Clean up stale entries (hands that disappeared)
        if now - self._last_cleanup > 2.0:
            self._last_cleanup = now
            stale = [k for k, (_, t) in self._finger_z.items() if now - t > 1.0]
            for k in stale:
                del self._finger_z[k]
                self._tap_cooldown.pop(k, None)

        return taps

    def destroy(self):
        if self.landmarker:
            self.landmarker.close()
